src/quicklist.c

```diff
@@ -1283,27 +1456,54 @@ int quicklistIndex(const quicklist *quicklist, const long long idx,
     }
 
     quicklistDecompressNodeForUse(entry->node);
-    entry->zi = ziplistIndex(entry->node->zl, entry->offset);
-    if (!ziplistGet(entry->zi, &entry->value, &entry->sz, &entry->longval))
+
+    if (unlikely(entry->node->container == QUICKLIST_NODE_CONTAINER_NONE)) {
+        entry->value = entry->node->entry;
+        entry->sz = entry->node->sz;
+        return 1;
+    }
+
+    entry->zi = ziplistIndex(entry->node->entry, entry->offset);
+    unsigned int sz = 0;
+    if (!ziplistGet(entry->zi, &entry->value, &sz, &entry->longval))
         assert(0); /* This can happen on corrupt ziplist with fake entry count. */
     /* The caller will use our result, so we don't re-compress here.
      * The caller can recompress or delete the node as needed. */
+    entry->sz = sz;
     return 1;
 }
 
 /* Rotate quicklist by moving the tail element to the head. */
 void quicklistRotate(quicklist *quicklist) {
     if (quicklist->count <= 1)
         return;
-
     /* First, get the tail entry */
-    unsigned char *p = ziplistIndex(quicklist->tail->zl, -1);
+    unsigned char *p = NULL;
     unsigned char *value, *tmp;
     long long longval;
-    unsigned int sz;
+    size_t sz;
     char longstr[32] = {0};
-    ziplistGet(p, &tmp, &sz, &longval);
 
+    if(unlikely(quicklist->tail->container == QUICKLIST_NODE_CONTAINER_NONE)) {
+
+        quicklistNode *head = quicklist->tail;
+        quicklistNode *tail = quicklist->head;
+
+        head->next = quicklist->head->next;
+        head->prev = NULL;
+
+        tail->next = NULL;
+        tail->prev = quicklist->tail->prev;
+
+        quicklist->head = head;
+        quicklist->tail = tail;
+
+        return;
+    }
```

Comment:
I think you misunderstood what the `quicklistRotate` method does.
top comment: `/* Rotate quicklist by moving the tail element to the head.*/`
It only moves the last element to the top, not the head to the tail.
Look at the code you have also moved the head to the tail.

Reply:
I think we should add a big element related test to REDIS_TEST.